### PR TITLE
tests: improve the deterministic behavior of the test suite

### PR DESCRIPTION
**Pull request by nviennot**
Various I/O objects are unclosed when the object falls out of scope.
This can lead to non-deterministic behavior.

### CODE REVIEW

**Comment by rst0git** [comment]
I think we need to update `__get_start_cmd()` as well. `wait()` may result in a [deadlock](https://docs.python.org/3/library/subprocess.html#subprocess.Popen.wait) when using `stdout=PIPE`.

File: test/zdtm.py [hunk 1]
```diff
@@ -202,6 +202,8 @@ def __copy_one(self, fname):
 
     def __copy_libs(self, binary):
         ldd = subprocess.Popen(["ldd", binary], stdout=subprocess.PIPE)
+        stdout, stderr = ldd.communicate()
```
Comment (rst0git) [inline test/zdtm.py:205]: `stderr` is not used, we could use `unused` or `_` instead
https://www.python.org/dev/peps/pep-0640/

File: test/zdtm.py [hunk 2]
```diff
@@ -331,8 +331,7 @@ def decode_flav(i):
 
 def tail(path):
     p = subprocess.Popen(['tail', '-n1', path], stdout=subprocess.PIPE)
-    out = p.stdout.readline()
-    p.wait()
+    out, err = p.communicate()
```
Comment (rst0git) [inline test/zdtm.py:334]: `err` is not used, we could use `unused` or `_` instead

File: test/zdtm.py (hunk 1 above)
Comment (nviennot) [reply test/zdtm.py]: Fixed

File: test/zdtm.py (hunk 2 above)
Comment (nviennot) [reply test/zdtm.py]: Fixed

**Comment by nviennot** [comment]
Regarding `__get_start_cmd()`, I also changed it to use `communicate()`. I had to fix a bit of the code to make things work (invoking `list()`).

Note that there was no risk of deadlock as there's only one output that is redirected (the program was already reading stdout correctly). There's a risk of deadlock when both stdout and stderr are redirected. Reading stdout and then stderr sequentially can create a deadlock as the child process may want to write to both stdout and stderr, and one of the pipe can fill up. communicate() drain both at the same time.

It's just good practice to use communicate. Also it waits for the process at the same time. It's nice.

**Comment by adrianreber** [comment]
Rebasing should remove most of the CI errors.

**Comment by nviennot** [comment]
Rebased

**Review by rst0git — APPROVED** [review]
LGTM

**Comment by adrianreber** [comment]
Another rebase is necessary. Looks reasonable.

**Comment by avagin** [comment]
LGTM

**Comment by nviennot** [comment]
Rebased

**Pull request merged**